File: src/LTAdmin/Repositories/BulletinRepository.cs
using System.Data;
using LTAdmin.Data;
using LTAdmin.Models.Dto;
using LTAdmin.Models.Entities;

namespace LTAdmin.Repositories;

/// <summary>Dépôt des bulletins, lignes, mentions et résultats finaux.</summary>
public sealed class BulletinRepository : RepositoryBase
{
    public BulletinRepository(AccessDatabase database) : base(database)
    {
    }

    // ----- BULLETIN -----

    public Bulletin? GetByInscriptionPeriode(int idInscription, int idPeriode)
        => QuerySingle(
            $"SELECT * FROM {Q(Tables.Bulletin)} WHERE {Q("ID_INSCRIPTION")} = ? AND {Q("ID_PERIODE")} = ?",
            MapBulletin, P(idInscription), P(idPeriode));

    public List<Bulletin> ListByClassePeriode(int idClasse, int idPeriode)
        => QueryList(
            $"SELECT B.* FROM {Q(Tables.Bulletin)} AS B " +
            $"INNER JOIN {Q(Tables.Inscription)} AS I ON B.{Q("ID_INSCRIPTION")} = I.{Q("ID_INSCRIPTION")} " +
            $"WHERE I.{Q("ID_CLASSE")} = ? AND B.{Q("ID_PERIODE")} = ? ORDER BY B.{Q("RANG")}",
            MapBulletin, P(idClasse), P(idPeriode));

    public List<BulletinResume> ListResumes(int? idClasse = null, int? idPeriode = null)
    {
        var conditions = new List<string>();
        var parameters = new List<System.Data.OleDb.OleDbParameter>();
        if (idClasse.HasValue) { conditions.Add($"I.{Q("ID_CLASSE")} = ?"); parameters.Add(P(idClasse.Value)); }
        if (idPeriode.HasValue) { conditions.Add($"B.{Q("ID_PERIODE")} = ?"); parameters.Add(P(idPeriode.Value)); }
        var sql = BulletinResumeSql();
        if (conditions.Count > 0) sql += " WHERE " + string.Join(" AND ", conditions);
        sql += $" ORDER BY C.{Q("LIBELLE")}, B.{Q("RANG")}";
        return QueryList(sql, MapBulletinResume, parameters.ToArray());
    }

    private static string BulletinResumeSql()
        => $"SELECT B.*, E.{Q("MATRICULE")}, E.{Q("NOM")}, E.{Q("PRENOM")}, " +
           $"C.{Q("LIBELLE")} AS CLASSE_LIB, PE.{Q("LIBELLE")} AS PERIODE_LIB " +
           $"FROM (((({Q(Tables.Bulletin)} AS B " +
           $"INNER JOIN {Q(Tables.Inscription)} AS I ON B.{Q("ID_INSCRIPTION")} = I.{Q("ID_INSCRIPTION")}) " +
           $"INNER JOIN {Q(Tables.Etudiant)} AS E ON I.{Q("ID_ETUDIANT")} = E.{Q("ID_ETUDIANT")}) " +
           $"INNER JOIN {Q(Tables.Classe)} AS C ON I.{Q("ID_CLASSE")} = C.{Q("ID_CLASSE")}) " +
           $"INNER JOIN {Q(Tables.PeriodeEval)} AS PE ON B.{Q("ID_PERIODE")} = PE.{Q("ID_PERIODE")})";

    public List<double> ListMoyennesByInscription(int idInscription)
    {
        var table = QueryTable(
            $"SELECT {Q("MOYENNE")} FROM {Q(Tables.Bulletin)} WHERE {Q("ID_INSCRIPTION")} = ? AND {Q("MOYENNE")} IS NOT NULL",
            P(idInscription));
        var result = new List<double>(table.Rows.Count);
        foreach (DataRow row in table.Rows)
        {
            var value = DataRowMapper.GetDouble(row, "MOYENNE");
            if (value.HasValue) result.Add(value.Value);
        }
        return result;
    }

    public int InsertBulletin(Bulletin bulletin)
        => InsertAndGetId(
            $"INSERT INTO {Q(Tables.Bulletin)} ({Q("ID_INSCRIPTION")}, {Q("ID_PERIODE")}, {Q("MOYENNE")}, " +
            $"{Q("TOTAL_POINTS")}, {Q("TOTAL_COEF")}, {Q("RANG")}, {Q("EFFECTIF")}, {Q("MOY_CLASSE")}, " +
            $"{Q("NB_ABSENCE")}, {Q("APPRECIATION")}, {Q("DECISION")}, {Q("DATE_EDITION")}) " +
            "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
            P(bulletin.IdInscription), P(bulletin.IdPeriode), P(bulletin.Moyenne),
            P(bulletin.TotalPoints), P(bulletin.TotalCoef), P(bulletin.Rang), P(bulletin.Effectif),
            P(bulletin.MoyClasse), P(bulletin.NbAbsence), P(bulletin.Appreciation),
            P(bulletin.Decision), P(bulletin.DateEdition));

    public int UpdateRang(Bulletin bulletin)
        => Execute(
            $"UPDATE {Q(Tables.Bulletin)} SET {Q("RANG")} = ?, {Q("EFFECTIF")} = ?, {Q("MOY_CLASSE")} = ? WHERE {Q("ID_BULLETIN")} = ?",
            P(bulletin.Rang), P(bulletin.Effectif), P(bulletin.MoyClasse), P(bulletin.IdBulletin));

    /// <summary>Identifiants des bulletins existants pour régénération ciblée.</summary>
    public List<int> ListIdsByClassePeriode(int idClasse, int idPeriode)
    {
        var table = QueryTable(
            $"SELECT B.{Q("ID_BULLETIN")} FROM {Q(Tables.Bulletin)} AS B " +
            $"INNER JOIN {Q(Tables.Inscription)} AS I ON B.{Q("ID_INSCRIPTION")} = I.{Q("ID_INSCRIPTION")} " +
            $"WHERE I.{Q("ID_CLASSE")} = ? AND B.{Q("ID_PERIODE")} = ?",
            P(idClasse), P(idPeriode));
        var result = new List<int>(table.Rows.Count);
        foreach (DataRow row in table.Rows)
        {
            var value = DataRowMapper.GetInt32(row, "ID_BULLETIN");
            if (value.HasValue) result.Add(value.Value);
        }
        return result;
    }

    public int DeleteBulletin(int idBulletin)
        => Execute($"DELETE FROM {Q(Tables.Bulletin)} WHERE {Q("ID_BULLETIN")} = ?", P(idBulletin));

    // ----- BULLETIN_LIGNE -----

    public List<BulletinLigne> ListLignes(int idBulletin)
        => QueryList(
            $"SELECT * FROM {Q(Tables.BulletinLigne)} WHERE {Q("ID_BULLETIN")} = ?",
            MapLigne, P(idBulletin));

    public int InsertLigne(BulletinLigne ligne)
        => InsertAndGetId(
            $"INSERT INTO {Q(Tables.BulletinLigne)} ({Q("ID_BULLETIN")}, {Q("CODE_MATIERE")}, {Q("MOYENNE_MAT")}, " +
            $"{Q("COEFFICIENT")}, {Q("POINTS")}, {Q("RANG_MAT")}, {Q("MOY_MIN")}, {Q("MOY_MAX")}, " +
            $"{Q("APPRECIATION")}, {Q("ID_FORMATEUR")}) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
            P(ligne.IdBulletin), P(ligne.CodeMatiere), P(ligne.MoyenneMat),
            P(ligne.Coefficient), P(ligne.Points), P(ligne.RangMat), P(ligne.MoyMin), P(ligne.MoyMax),
            P(ligne.Appreciation), P(ligne.IdFormateur));

    public int UpdateLigneStats(BulletinLigne ligne)
        => Execute(
            $"UPDATE {Q(Tables.BulletinLigne)} SET {Q("RANG_MAT")} = ?, {Q("MOY_MIN")} = ?, {Q("MOY_MAX")} = ? WHERE {Q("ID_LIGNE")} = ?",
            P(ligne.RangMat), P(ligne.MoyMin), P(ligne.MoyMax), P(ligne.IdLigne));

    public int DeleteLignesByBulletin(int idBulletin)
        => Execute($"DELETE FROM {Q(Tables.BulletinLigne)} WHERE {Q("ID_BULLETIN")} = ?", P(idBulletin));

    // ----- GRILLE_MENTION -----

    public List<GrilleMention> ListMentions()
        => QueryList(
            $"SELECT * FROM {Q(Tables.GrilleMention)} ORDER BY {Q("INF")} DESC",
            MapMention);

    // ----- RESULTAT_FINAL -----

    public ResultatFinal? GetByInscription(int idInscription)
        => QuerySingle(
            $"SELECT * FROM {Q(Tables.ResultatFinal)} WHERE {Q("ID_INSCRIPTION")} = ?",
            MapResultat, P(idInscription));

    public List<ResultatFinal> ListByClasse(int idClasse)
        => QueryList(
            $"SELECT R.* FROM {Q(Tables.ResultatFinal)} AS R " +
            $"INNER JOIN {Q(Tables.Inscription)} AS I ON R.{Q("ID_INSCRIPTION")} = I.{Q("ID_INSCRIPTION")} " +
            $"WHERE I.{Q("ID_CLASSE")} = ? ORDER BY R.{Q("RANG")}",
            MapResultat, P(idClasse));

    public int InsertResultat(ResultatFinal resultat)
        => InsertAndGetId(
            $"INSERT INTO {Q(Tables.ResultatFinal)} ({Q("ID_INSCRIPTION")}, {Q("MOY_CC")}, {Q("MOY_EXAM")}, " +
            $"{Q("MOYENNE_GEN")}, {Q("RANG")}, {Q("MENTION")}, {Q("DECISION")}, {Q("CREDIT_VALIDE")}, " +
            $"{Q("DATE_DELIB")}, {Q("OBSERVATION")}) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
            P(resultat.IdInscription), P(resultat.MoyCc), P(resultat.MoyExam),
            P(resultat.MoyenneGen), P(resultat.Rang), P(resultat.Mention), P(resultat.Decision),
            P(resultat.CreditValide), P(resultat.DateDelib), P(resultat.Observation));

    public int UpdateResultatRang(ResultatFinal resultat)
        => Execute(
            $"UPDATE {Q(Tables.ResultatFinal)} SET {Q("RANG")} = ? WHERE {Q("ID_RESULTAT")} = ?",
            P(resultat.Rang), P(resultat.IdResultat));

    public int DeleteByInscription(int idInscription)
        => Execute($"DELETE FROM {Q(Tables.ResultatFinal)} WHERE {Q("ID_INSCRIPTION")} = ?", P(idInscription));

    // ----- Mappers -----

    public static Bulletin MapBulletin(DataRow row) => new()
    {
        IdBulletin = DataRowMapper.GetInt32(row, "ID_BULLETIN"),
        IdInscription = DataRowMapper.GetInt32(row, "ID_INSCRIPTION"),
        IdPeriode = DataRowMapper.GetInt32(row, "ID_PERIODE"),
        Moyenne = DataRowMapper.GetDouble(row, "MOYENNE"),
        TotalPoints = DataRowMapper.GetDouble(row, "TOTAL_POINTS"),
        TotalCoef = DataRowMapper.GetDouble(row, "TOTAL_COEF"),
        Rang = DataRowMapper.GetInt32(row, "RANG"),
        Effectif = DataRowMapper.GetInt32(row, "EFFECTIF"),
        MoyClasse = DataRowMapper.GetDouble(row, "MOY_CLASSE"),
        NbAbsence = DataRowMapper.GetDouble(row, "NB_ABSENCE"),
        Appreciation = DataRowMapper.GetString(row, "APPRECIATION"),
        Decision = DataRowMapper.GetString(row, "DECISION"),
        DateEdition = DataRowMapper.GetDateTime(row, "DATE_EDITION")
    };

    public static BulletinResume MapBulletinResume(DataRow row) => new()
    {
        IdBulletin = DataRowMapper.GetInt32(row, "ID_BULLETIN"),
        IdInscription = DataRowMapper.GetInt32(row, "ID_INSCRIPTION"),
        Matricule = DataRowMapper.GetString(row, "MATRICULE"),
        Nom = DataRowMapper.GetString(row, "NOM"),
        Prenom = DataRowMapper.GetString(row, "PRENOM"),
        Classe = DataRowMapper.GetString(row, "CLASSE_LIB"),
        IdPeriode = DataRowMapper.GetInt32(row, "ID_PERIODE"),
        Periode = DataRowMapper.GetString(row, "PERIODE_LIB"),
        Moyenne = DataRowMapper.GetDouble(row, "MOYENNE"),
        Rang = DataRowMapper.GetInt32(row, "RANG"),
        Effectif = DataRowMapper.GetInt32(row, "EFFECTIF"),
        Decision = DataRowMapper.GetString(row, "DECISION")
    };

    public static BulletinLigne MapLigne(DataRow row) => new()
    {
        IdLigne = DataRowMapper.GetInt32(row, "ID_LIGNE"),
        IdBulletin = DataRowMapper.GetInt32(row, "ID_BULLETIN"),
        CodeMatiere = DataRowMapper.GetString(row, "CODE_MATIERE"),
        MoyenneMat = DataRowMapper.GetDouble(row, "MOYENNE_MAT"),
        Coefficient = DataRowMapper.GetDouble(row, "COEFFICIENT"),
        Points = DataRowMapper.GetDouble(row, "POINTS"),
        RangMat = DataRowMapper.GetInt32(row, "RANG_MAT"),
        MoyMin = DataRowMapper.GetDouble(row, "MOY_MIN"),
        MoyMax = DataRowMapper.GetDouble(row, "MOY_MAX"),
        Appreciation = DataRowMapper.GetString(row, "APPRECIATION"),
        IdFormateur = DataRowMapper.GetInt32(row, "ID_FORMATEUR")
    };

    public static GrilleMention MapMention(DataRow row) => new()
    {
        IdMention = DataRowMapper.GetInt32(row, "ID_MENTION"),
        Inf = DataRowMapper.GetDouble(row, "INF"),
        Sup = DataRowMapper.GetDouble(row, "SUP"),
        Mention = DataRowMapper.GetString(row, "MENTION"),
        Admis = DataRowMapper.GetBoolean(row, "ADMIS")
    };

    public static ResultatFinal MapResultat(DataRow row) => new()
    {
        IdResultat = DataRowMapper.GetInt32(row, "ID_RESULTAT"),
        IdInscription = DataRowMapper.GetInt32(row, "ID_INSCRIPTION"),
        MoyCc = DataRowMapper.GetDouble(row, "MOY_CC"),
        MoyExam = DataRowMapper.GetDouble(row, "MOY_EXAM"),
        MoyenneGen = DataRowMapper.GetDouble(row, "MOYENNE_GEN"),
        Rang = DataRowMapper.GetInt32(row, "RANG"),
        Mention = DataRowMapper.GetString(row, "MENTION"),
        Decision = DataRowMapper.GetString(row, "DECISION"),
        CreditValide = DataRowMapper.GetInt32(row, "CREDIT_VALIDE"),
        DateDelib = DataRowMapper.GetDateTime(row, "DATE_DELIB"),
        Observation = DataRowMapper.GetString(row, "OBSERVATION")
    };
}

File: ltadmin/repositories/bulletin_repository.py
"""Dépôt des bulletins, lignes, mentions et résultats finaux."""

from __future__ import annotations

from typing import List, Optional

from ltadmin.data import row_mapper as rm
from ltadmin.data.schema import Tables
from ltadmin.models.dto import BulletinResume
from ltadmin.models.entities import Bulletin, BulletinLigne, GrilleMention, ResultatFinal
from ltadmin.repositories.base import RepositoryBase, p, q


class BulletinRepository(RepositoryBase):

    # ----- BULLETIN -----

    def get_by_inscription_periode(self, id_inscription: int, id_periode: int) -> Optional[Bulletin]:
        return self.query_single(
            f"SELECT * FROM {q(Tables.BULLETIN)} WHERE {q('ID_INSCRIPTION')} = ? "
            f"AND {q('ID_PERIODE')} = ?",
            BulletinRepository.map_bulletin, [p(id_inscription), p(id_periode)])

    def list_by_classe_periode(self, id_classe: int, id_periode: int) -> List[Bulletin]:
        return self.query_list(
            f"SELECT B.* FROM {q(Tables.BULLETIN)} AS B "
            f"INNER JOIN {q(Tables.INSCRIPTION)} AS I ON B.{q('ID_INSCRIPTION')} = "
            f"I.{q('ID_INSCRIPTION')} "
            f"WHERE I.{q('ID_CLASSE')} = ? AND B.{q('ID_PERIODE')} = ? ORDER BY B.{q('RANG')}",
            BulletinRepository.map_bulletin, [p(id_classe), p(id_periode)])

    def list_resumes(self, id_classe: Optional[int] = None,
                     id_periode: Optional[int] = None) -> List[BulletinResume]:
        conditions = []
        parameters = []
        if id_classe is not None:
            conditions.append(f"I.{q('ID_CLASSE')} = ?")
            parameters.append(p(id_classe))
        if id_periode is not None:
            conditions.append(f"B.{q('ID_PERIODE')} = ?")
            parameters.append(p(id_periode))
        sql = BulletinRepository.bulletin_resume_sql()
        if conditions:
            sql += " WHERE " + " AND ".join(conditions)
        sql += f" ORDER BY C.{q('LIBELLE')}, B.{q('RANG')}"
        return self.query_list(sql, BulletinRepository.map_bulletin_resume, parameters)

    @staticmethod
    def bulletin_resume_sql() -> str:
        return (
            f"SELECT B.*, E.{q('MATRICULE')}, E.{q('NOM')}, E.{q('PRENOM')}, "
            f"C.{q('LIBELLE')} AS CLASSE_LIB, PE.{q('LIBELLE')} AS PERIODE_LIB "
            f"FROM (((({q(Tables.BULLETIN)} AS B "
            f"INNER JOIN {q(Tables.INSCRIPTION)} AS I ON B.{q('ID_INSCRIPTION')} = "
            f"I.{q('ID_INSCRIPTION')}) "
            f"INNER JOIN {q(Tables.ETUDIANT)} AS E ON I.{q('ID_ETUDIANT')} = E.{q('ID_ETUDIANT')}) "
            f"INNER JOIN {q(Tables.CLASSE)} AS C ON I.{q('ID_CLASSE')} = C.{q('ID_CLASSE')}) "
            f"INNER JOIN {q(Tables.PERIODE_EVAL)} AS PE ON B.{q('ID_PERIODE')} = "
            f"PE.{q('ID_PERIODE')})"
        )

    def list_moyennes_by_inscription(self, id_inscription: int) -> List[float]:
        table = self.query_table(
            f"SELECT {q('MOYENNE')} FROM {q(Tables.BULLETIN)} "
            f"WHERE {q('ID_INSCRIPTION')} = ? AND {q('MOYENNE')} IS NOT NULL",
            [p(id_inscription)])
        result = []
        for row in table:
            value = rm.get_double(row, "MOYENNE")
            if value is not None:
                result.append(value)
        return result

    def insert_bulletin(self, bulletin: Bulletin) -> int:
        return self.insert_and_get_id(
            f"INSERT INTO {q(Tables.BULLETIN)} ({q('ID_INSCRIPTION')}, {q('ID_PERIODE')}, "
            f"{q('MOYENNE')}, {q('TOTAL_POINTS')}, {q('TOTAL_COEF')}, {q('RANG')}, "
            f"{q('EFFECTIF')}, {q('MOY_CLASSE')}, {q('NB_ABSENCE')}, {q('APPRECIATION')}, "
            f"{q('DECISION')}, {q('DATE_EDITION')}) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
            [p(bulletin.id_inscription), p(bulletin.id_periode), p(bulletin.moyenne),
             p(bulletin.total_points), p(bulletin.total_coef), p(bulletin.rang),
             p(bulletin.effectif), p(bulletin.moy_classe), p(bulletin.nb_absence),
             p(bulletin.appreciation), p(bulletin.decision), p(bulletin.date_edition)])

    def update_rang(self, bulletin: Bulletin) -> int:
        return self.execute(
            f"UPDATE {q(Tables.BULLETIN)} SET {q('RANG')} = ?, {q('EFFECTIF')} = ?, "
            f"{q('MOY_CLASSE')} = ? WHERE {q('ID_BULLETIN')} = ?",
            [p(bulletin.rang), p(bulletin.effectif), p(bulletin.moy_classe),
             p(bulletin.id_bulletin)])

    def update_appreciation(self, bulletin: Bulletin) -> int:
        """Modifie l'appréciation générale d'un bulletin (sans recalcul)."""
        return self.execute(
            f"UPDATE {q(Tables.BULLETIN)} SET {q('APPRECIATION')} = ? "
            f"WHERE {q('ID_BULLETIN')} = ?",
            [p(bulletin.appreciation), p(bulletin.id_bulletin)])

    def list_ids_by_classe_periode(self, id_classe: int, id_periode: int) -> List[int]:
        """Identifiants des bulletins existants pour régénération ciblée."""
        table = self.query_table(
            f"SELECT B.{q('ID_BULLETIN')} FROM {q(Tables.BULLETIN)} AS B "
            f"INNER JOIN {q(Tables.INSCRIPTION)} AS I ON B.{q('ID_INSCRIPTION')} = "
            f"I.{q('ID_INSCRIPTION')} "
            f"WHERE I.{q('ID_CLASSE')} = ? AND B.{q('ID_PERIODE')} = ?",
            [p(id_classe), p(id_periode)])
        result = []
        for row in table:
            value = rm.get_int(row, "ID_BULLETIN")
            if value is not None:
                result.append(value)
        return result

    def delete_bulletin(self, id_bulletin: int) -> int:
        return self.execute(
            f"DELETE FROM {q(Tables.BULLETIN)} WHERE {q('ID_BULLETIN')} = ?", [p(id_bulletin)])

    # ----- BULLETIN_LIGNE -----

    def list_lignes(self, id_bulletin: int) -> List[BulletinLigne]:
        return self.query_list(
            f"SELECT * FROM {q(Tables.BULLETIN_LIGNE)} WHERE {q('ID_BULLETIN')} = ?",
            BulletinRepository.map_ligne, [p(id_bulletin)])

    def insert_ligne(self, ligne: BulletinLigne) -> int:
        return self.insert_and_get_id(
            f"INSERT INTO {q(Tables.BULLETIN_LIGNE)} ({q('ID_BULLETIN')}, {q('CODE_MATIERE')}, "
            f"{q('MOYENNE_MAT')}, {q('COEFFICIENT')}, {q('POINTS')}, {q('RANG_MAT')}, "
            f"{q('MOY_MIN')}, {q('MOY_MAX')}, {q('APPRECIATION')}, {q('ID_FORMATEUR')}) "
            f"VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
            [p(ligne.id_bulletin), p(ligne.code_matiere), p(ligne.moyenne_mat),
             p(ligne.coefficient), p(ligne.points), p(ligne.rang_mat), p(ligne.moy_min),
             p(ligne.moy_max), p(ligne.appreciation), p(ligne.id_formateur)])

    def update_ligne_stats(self, ligne: BulletinLigne) -> int:
        return self.execute(
            f"UPDATE {q(Tables.BULLETIN_LIGNE)} SET {q('RANG_MAT')} = ?, {q('MOY_MIN')} = ?, "
            f"{q('MOY_MAX')} = ? WHERE {q('ID_LIGNE')} = ?",
            [p(ligne.rang_mat), p(ligne.moy_min), p(ligne.moy_max), p(ligne.id_ligne)])

    def delete_lignes_by_bulletin(self, id_bulletin: int) -> int:
        return self.execute(
            f"DELETE FROM {q(Tables.BULLETIN_LIGNE)} WHERE {q('ID_BULLETIN')} = ?",
            [p(id_bulletin)])

    # ----- GRILLE_MENTION -----

    def list_mentions(self) -> List[GrilleMention]:
        return self.query_list(
            f"SELECT * FROM {q(Tables.GRILLE_MENTION)} ORDER BY {q('INF')} DESC",
            BulletinRepository.map_mention)

    # ----- RESULTAT_FINAL -----

    def get_by_inscription(self, id_inscription: int) -> Optional[ResultatFinal]:
        return self.query_single(
            f"SELECT * FROM {q(Tables.RESULTAT_FINAL)} WHERE {q('ID_INSCRIPTION')} = ?",
            BulletinRepository.map_resultat, [p(id_inscription)])

    def list_by_classe(self, id_classe: int) -> List[ResultatFinal]:
        return self.query_list(
            f"SELECT R.* FROM {q(Tables.RESULTAT_FINAL)} AS R "
            f"INNER JOIN {q(Tables.INSCRIPTION)} AS I ON R.{q('ID_INSCRIPTION')} = "
            f"I.{q('ID_INSCRIPTION')} "
            f"WHERE I.{q('ID_CLASSE')} = ? ORDER BY R.{q('RANG')}",
            BulletinRepository.map_resultat, [p(id_classe)])

    def insert_resultat(self, resultat: ResultatFinal) -> int:
        return self.insert_and_get_id(
            f"INSERT INTO {q(Tables.RESULTAT_FINAL)} ({q('ID_INSCRIPTION')}, {q('MOY_CC')}, "
            f"{q('MOY_EXAM')}, {q('MOYENNE_GEN')}, {q('RANG')}, {q('MENTION')}, "
            f"{q('DECISION')}, {q('CREDIT_VALIDE')}, {q('DATE_DELIB')}, {q('OBSERVATION')}) "
            f"VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
            [p(resultat.id_inscription), p(resultat.moy_cc), p(resultat.moy_exam),
             p(resultat.moyenne_gen), p(resultat.rang), p(resultat.mention),
             p(resultat.decision), p(resultat.credit_valide), p(resultat.date_delib),
             p(resultat.observation)])

    def update_resultat_rang(self, resultat: ResultatFinal) -> int:
        return self.execute(
            f"UPDATE {q(Tables.RESULTAT_FINAL)} SET {q('RANG')} = ? WHERE {q('ID_RESULTAT')} = ?",
            [p(resultat.rang), p(resultat.id_resultat)])

    def delete_by_inscription(self, id_inscription: int) -> int:
        return self.execute(
            f"DELETE FROM {q(Tables.RESULTAT_FINAL)} WHERE {q('ID_INSCRIPTION')} = ?",
            [p(id_inscription)])

    # ----- Mappers -----

    @staticmethod
    def map_bulletin(row: dict) -> Bulletin:
        return Bulletin(
            id_bulletin=rm.get_int(row, "ID_BULLETIN"),
            id_inscription=rm.get_int(row, "ID_INSCRIPTION"),
            id_periode=rm.get_int(row, "ID_PERIODE"),
            moyenne=rm.get_double(row, "MOYENNE"),
            total_points=rm.get_double(row, "TOTAL_POINTS"),
            total_coef=rm.get_double(row, "TOTAL_COEF"),
            rang=rm.get_int(row, "RANG"),
            effectif=rm.get_int(row, "EFFECTIF"),
            moy_classe=rm.get_double(row, "MOY_CLASSE"),
            nb_absence=rm.get_double(row, "NB_ABSENCE"),
            appreciation=rm.get_string(row, "APPRECIATION"),
            decision=rm.get_string(row, "DECISION"),
            date_edition=rm.get_datetime(row, "DATE_EDITION"),
        )

    @staticmethod
    def map_bulletin_resume(row: dict) -> BulletinResume:
        return BulletinResume(
            id_bulletin=rm.get_int(row, "ID_BULLETIN"),
            id_inscription=rm.get_int(row, "ID_INSCRIPTION"),
            matricule=rm.get_string(row, "MATRICULE"),
            nom=rm.get_string(row, "NOM"),
            prenom=rm.get_string(row, "PRENOM"),
            classe=rm.get_string(row, "CLASSE_LIB"),
            id_periode=rm.get_int(row, "ID_PERIODE"),
            periode=rm.get_string(row, "PERIODE_LIB"),
            moyenne=rm.get_double(row, "MOYENNE"),
            rang=rm.get_int(row, "RANG"),
            effectif=rm.get_int(row, "EFFECTIF"),
            decision=rm.get_string(row, "DECISION"),
        )

    @staticmethod
    def map_ligne(row: dict) -> BulletinLigne:
        return BulletinLigne(
            id_ligne=rm.get_int(row, "ID_LIGNE"),
            id_bulletin=rm.get_int(row, "ID_BULLETIN"),
            code_matiere=rm.get_string(row, "CODE_MATIERE"),
            moyenne_mat=rm.get_double(row, "MOYENNE_MAT"),
            coefficient=rm.get_double(row, "COEFFICIENT"),
            points=rm.get_double(row, "POINTS"),
            rang_mat=rm.get_int(row, "RANG_MAT"),
            moy_min=rm.get_double(row, "MOY_MIN"),
            moy_max=rm.get_double(row, "MOY_MAX"),
            appreciation=rm.get_string(row, "APPRECIATION"),
            id_formateur=rm.get_int(row, "ID_FORMATEUR"),
        )

    @staticmethod
    def map_mention(row: dict) -> GrilleMention:
        return GrilleMention(
            id_mention=rm.get_int(row, "ID_MENTION"),
            inf=rm.get_double(row, "INF"),
            sup=rm.get_double(row, "SUP"),
            mention=rm.get_string(row, "MENTION"),
            admis=rm.get_bool(row, "ADMIS"),
        )

    @staticmethod
    def map_resultat(row: dict) -> ResultatFinal:
        return ResultatFinal(
            id_resultat=rm.get_int(row, "ID_RESULTAT"),
            id_inscription=rm.get_int(row, "ID_INSCRIPTION"),
            moy_cc=rm.get_double(row, "MOY_CC"),
            moy_exam=rm.get_double(row, "MOY_EXAM"),
            moyenne_gen=rm.get_double(row, "MOYENNE_GEN"),
            rang=rm.get_int(row, "RANG"),
            mention=rm.get_string(row, "MENTION"),
            decision=rm.get_string(row, "DECISION"),
            credit_valide=rm.get_int(row, "CREDIT_VALIDE"),
            date_delib=rm.get_datetime(row, "DATE_DELIB"),
            observation=rm.get_string(row, "OBSERVATION"),
        )
